Consolidate Address types

alexflorisca opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
Presently we have 2 different types for a user entered address. This PR consolidates the 2 types into one and updates its use throughout the code

Fixes #6627

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

Commit:
Consolidate Address types

Commit:
Merge branch 'trunk' into update/update-address-types

opr commented on assets/js/types/type-defs/addresses.ts:
```diff
@@ -1,13 +1,13 @@
 	postcode: string;
-	state?: string;
+	phone?: string;
+	email?: string;
```
I'm wondering about whether it is correct to use this type for the shipping address, email is not a field on a shipping address, but here it is marked as optional. So if we're writing code, our IDE will tell us `shippingAddress.email` is fine to do, but it actually isn't. I think we should still try to split them up into distinct address types if possible.

opr reviewed (changes requested):
Hey @alexflorisca thanks for this work! I would prefer to not use `EnteredAddress` it seems like it's too specific/not descriptive enough for what it actually represents. Could we perhaps try `CustomerAddress` or similar?

Commit:
Revert "Consolidate Address types"

This reverts commit cb708a3d714621717414ec17f5e1071be5ca5e96.

Commit:
Rename EnteredAddress to CustomerAddress

Commit:
Delete unused Address type

alexflorisca commented:
Yep good call, I went the other way with this PR now. I reverted the original commit, Renamed `EnteredAddress` to `CustomerAddress` and removed the `Address` type from /type-defs

Commit:
Merge branch 'trunk' into update/update-address-types

opr commented on assets/js/types/type-defs/checkout.ts:
```diff
@@ -16,7 +16,7 @@ export interface CheckoutResponseSuccess {
 		redirect_url: string;
 	};
-	shipping_address: Address;
+	shipping_address: CustomerAddress;
```
Should this be `ShippingAddress` (I know this is just an alias of `CustomerAddress` but would make it easier to reason with)

opr commented on assets/js/data/checkout/constants.ts:
```diff
@@ -30,7 +30,7 @@ const preloadedCheckoutData = getSetting(
 	customer_id: 0,
-	billing_address: {} as EnteredAddress,
-	shipping_address: {} as EnteredAddress,
+	billing_address: {} as CustomerAddress,
```
Should these be `BillingAddress` and `ShippingAddress`?

opr commented on assets/js/base/utils/address.ts:
```diff
@@ -7,19 +7,19 @@ import type {
-			address1[ field as keyof EnteredAddress ] ===
-			address2[ field as keyof EnteredAddress ]
+			address1[ field as keyof CustomerAddress ] ===
+			address2[ field as keyof CustomerAddress ]
```
We could make this generic:

```ts
import {
	defaultAddressFields,
	ShippingAddress,
	BillingAddress,
} from '@woocommerce/settings';

export const isSameAddress = < T extends ShippingAddress | BillingAddress >(
	address1: T,
	address2: T
): boolean => {
	return Object.keys( defaultAddressFields ).every(
		( field: string ) =>
			address1[ field as keyof T ] === address2[ field as keyof T ]
	);
};
```

alexflorisca replied:
Yeah nice suggestion 🚀 !

opr commented on assets/js/base/context/hooks/use-checkout-address.ts:
```diff
@@ -21,8 +21,8 @@ import { useShippingData } from './shipping/use-shipping-data';
-	setShippingAddress: ( data: Partial< EnteredAddress > ) => void;
-	setBillingAddress: ( data: Partial< EnteredAddress > ) => void;
+	setShippingAddress: ( data: Partial< CustomerAddress > ) => void;
+	setBillingAddress: ( data: Partial< CustomerAddress > ) => void;
```
Should this be `BillingAddress` and `ShippingAddress`?

If you try to do `setBillingAddress( { email: 'test@test.com' } )` you get 

`Argument of type '{ email: string; }' is not assignable to parameter of type 'Partial<CustomerAddress>'.   Object literal may only specify known properties, and 'email' does not exist in type 'Partial<CustomerAddress>'.`

opr commented on assets/js/settings/shared/default-address-fields.ts:
```diff
@@ -54,8 +54,8 @@ export type KeyedAddressField = AddressField & {
-export type ShippingAddress = EnteredAddress;
-export interface BillingAddress extends EnteredAddress {
+export type ShippingAddress = CustomerAddress;
+export interface BillingAddress extends CustomerAddress {
```
Do we actually need `CustomerAddress` since it's instantly used to create `ShippingAddress`. I think we could rename it to `ShippingAddress`, remove the current `export type ShippingAddress = CustomerAddress;` line and `BillingAddress` extend `ShippingAddress`. 

Does this make sense?

opr reviewed (changes requested):
Thanks @alexflorisca this looks better, but I've left a few comments about `CustomerAddress` and a few cases when we should be more specific about which type of address we're dealing with.

Let me know what you think!

Commit:
Thomas suggestions

alexflorisca commented:
Nice suggestions man, much prish 🙏 . I've implemented all of them

Commit:
Merge branch 'trunk' into update/update-address-types

opr reviewed (approved):
Nice one this looks great!

Commit:
Merge branch 'trunk' into update/update-address-types

Commit:
Fix duplicate ShippingAddress symbol

Commit:
Merge branch 'update/update-address-types' of github.com:woocommerce/woocommerce-blocks into update/update-address-types
